src/Components/Blazor/Build/src/Tasks/GenerateBlazorBootJson.cs

```diff
@@ -27,60 +29,66 @@ public class GenerateBlazorBootJson : Task
-
-            using var fileStream = File.Create(OutputPath);
-            WriteBootJson(fileStream, entryAssemblyName, assemblies, LinkerEnabled);
-
-            return true;
-
-            static string GetUriPath(ITaskItem item)
+            // Build a two-level dictionary of the form:
+            // - BootResourceType (e.g., "assembly")
+            //   - UriPath (e.g., "System.Text.Json.dll")
+            //     - ContentHash (e.g., "4548fa2e9cf52986")
+            var resourcesByGroup = new Dictionary<string, Dictionary<string, string>>();
+            foreach (var resource in Resources)
             {
-                var outputPath = item.GetMetadata("RelativeOutputPath");
-                if (string.IsNullOrEmpty(outputPath))
+                var resourceType = resource.GetMetadata("BootResourceType");
+                if (string.IsNullOrEmpty(resourceType))
                 {
-                    outputPath = Path.GetFileName(item.ItemSpec);
+                    continue;
                 }
 
-                return outputPath.Replace('\\', '/');
+                if (!resourcesByGroup.TryGetValue(resourceType, out var group))
+                {
+                    group = new Dictionary<string, string>();
+                    resourcesByGroup.Add(resourceType, group);
+                }
+
+                var uriPath = GetUriPath(resource);
+                if (!group.ContainsKey(uriPath))
+                {
+                    // It's safe to truncate to a fairly short string, since the hash is not used for any
+                    // security purpose - the developer produces these files themselves, and the hash is
+                    // only used to check whether an earlier cached copy is up-to-date.
+                    // This truncation halves the size of blazor.boot.json in typical cases.
+                    group.Add(uriPath, resource.GetMetadata("FileHash").Substring(0, 16).ToLowerInvariant());
+                }
             }
-        }
 
-        internal static void WriteBootJson(Stream stream, string entryAssemblyName, string[] assemblies, bool linkerEnabled)
-        {
-            var data = new BootJsonData
+            var bootJsonData = new
             {
-                entryAssembly = entryAssemblyName,
-                assemblies = assemblies,
-                linkerEnabled = linkerEnabled,
+                EntryAssembly = AssemblyName.GetAssemblyName(AssemblyPath).Name,
+                Resources = resourcesByGroup,
+                DebugBuild,
+                LinkerEnabled,
             };
 
-            var serializer = new DataContractJsonSerializer(typeof(BootJsonData));
-            serializer.WriteObject(stream, data);
+            using (var fileStream = File.Create(OutputPath))
+            using (var utf8Writer = new Utf8JsonWriter(fileStream, new JsonWriterOptions { Indented = true }))
+            {
+                JsonSerializer.Serialize(utf8Writer, bootJsonData, new JsonSerializerOptions
```

Comment:
I'm not sure that this is safe to do unfortunately. @nguerrera

Reply:
Could you clarify what aspect of this may be unsafe?

Reply:
Referencing S.T.J in an MSBuild task. If anyone else in the entire universe references a different version of S.T.J in a build task this can blow up. Whichever task loads first will get to load its desired version, and the other task might lose if it requires a higher version.

Basically it's the problem that Binding Redirects solve, but there's no solution for MSBuild.

Reply:
I see - thanks! I’ll switch it back to DataContractJsonSerializer then.

Reply:
I think we'll manage to resolve this in 5.0, at least for components that version exactly with the SDK, but I know you're target is earlier and OOB, right?

Reply:
Switched back to `DataContractJsonSerializer`.